File: AdventOfCode.FSharp/2022/Day3.fs
﻿namespace AdventOfCode.FSharp._2022

open AdventOfCode.Core
open AdventOfCode.FSharp
open Xunit

module Day3 =
    let priorities = seq { seq { 'a' .. 'z' } ; seq { 'A' .. 'Z' } } |> Seq.concat |> Seq.mapi (fun i x -> (x, 1 + i)) |> dict

    let Solve = 
        Shared.Split 
        >> Seq.map (fun s -> Seq.find (fun (ss : char) -> s[(s.Length / 2)..(s.Length)].Contains(ss)) s[0..(s.Length / 2)])
        >> Seq.map (fun c -> priorities[c])
        >> Seq.sum

    let exampleInput =
        @"vJrwpWtwJgWrhcsFMMfFFhFp
jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
PmmdzqPrVvPwwTWBwg
wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
ttgJtRGJQctTZtZT
CrZsJsPPZsGzwwsLwLmpwMDw"

    [<Fact>]
    let ``Solves Part One Example`` () =
        let actual = Solve exampleInput
        Assert.Equal(157, actual)
        

type Day3(input: string) =
    interface ISolver with
        member this.SolvePartOne() = Day3.Solve input |> string
        member this.SolvePartTwo() = failwith "Solve part 1 first"

File: AdventOfCode.CSharp/2023/Day3.cs
using System.Text;
using Xunit;
using Xunit.Abstractions;

namespace AdventOfCode.CSharp._2023;

public class Day3 : Solver {
    public Day3(string? input = null, ITestOutputHelper? outputHelper = null) : base(input, outputHelper) { }

    public override long SolvePartOne() {
        var lines = Shared.Split(Input).Select(l => l.ToCharArray()).ToArray();

        var h = lines.Length + 2;
        var w = lines[0].Length + 2;
        var schematic = new char[h, w];

        for (var y = 0; y < h; ++y)
        for (var x = 0; x < w; ++x)
            if (y == 0 || x == 0 || y == h - 1 || x == w - 1)
                schematic[y, x] = '.';
            else
                schematic[y, x] = lines[y - 1][x - 1];

        var buffer = new StringBuilder(w);
        var accumulator = 0;
        
        for (var y = 0; y < h; ++y) {
            buffer.Clear();
                
            for (var x = 0; x < w; ++x) {
                if (char.IsNumber(schematic[y, x])) {
                    buffer.Append(schematic[y, x]);
                    continue;
                }

                if (buffer.Length > 0) {
                    var value = int.Parse(buffer.ToString());
                    //Output.WriteLine(value);

                    var isSymbol = false;
                    isSymbol |= schematic[y, x] != '.'; // right-inline
                    isSymbol |= schematic[y, x - buffer.Length - 1] != '.'; // left-inline
                    for (var ix = x - buffer.Length - 1; ix <= x; ++ix) {
                        isSymbol |= schematic[y-1, ix] != '.'; // aboxe
                        isSymbol |= schematic[y+1, ix] != '.'; // below
                    }
                    
                    //Output.WriteLine(isSymbol);
                    if (isSymbol) accumulator += value;
                    buffer.Clear();
                }
            }
        }

        return accumulator;
    }

    public override long SolvePartTwo() => throw new NotImplementedException("Solve part 1 first");

    private const string? ExampleInput = @"
467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..
";

    [Fact]
    public void SolvesPartOneExample() {
        var actual = new Day3(ExampleInput, Output).SolvePartOne();
        Assert.Equal(4361, actual);
    }
}